keps/sig-testing/0014-20180523-test-cluster-framework.md

```diff
@@ -0,0 +1,169 @@
+# Test Cluster Framework
+
+## Table of Contents
+
+* [Test Cluster Framework](#test-cluster-framework)
+   * [Table of Contents](#table-of-contents)
+   * [Summary](#summary)
+   * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+   * [Proposal](#proposal)
+      * [User Stories](#user-stories)
+      * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Risks and Mitigations](#risks-and-mitigations)
+   * [Graduation Criteria](#graduation-criteria)
+   * [Implementation History](#implementation-history)
+
+<!--
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+-->
+
+## Summary
+
+> Making testing with clusters consistent
+
+We propose an abstraction so that users and tools (e.g. CI) have a consistent
+way to spin up and configure clusters they can test against. This abstraction
+is targeted mainly to the needs of the following personas and should work
+equally well used interactively and non-interactively / automated.
+
+Alice, an experienced CLI Dev (see [Key Persona Alice][alice]), finds that she can’t
+easily onboard new contributors like Bob (see [Key Persona Bob][bob]), in part
+because her CLI and tests are too tightly coupled with heavyweight
+infrastructure in `k8s.io/kubernetes`. To solve this problem, Alice sponsors
+the development of a new lightweight testing infrastructure outside
+`k8s.io/kubernetes`.
+
+Meanwhile, Carol, an API/Controller dev (see [Key Persona Carol][carol]), is
+frustrated that there are already too many ways to spin up test clusters.
+
+There is a clear tension between Alice and Carol’s frustrations. Left to their
+own devices, Alice is likely to contribute to the proliferation of test
+infrastructure that frustrates Carol, and Carol may find herself advocating
+that Alice be prevented from acting to resolve her own frustration with the
+status quo.
+
+Most of the content of this KEP has already [been worked on and fleshed
+out][the_doc] mostly by people from sig-testing, more specifically people
+working on the [testing-commons subproject][testing_commons_notes].
+
+[key_personas]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.czu8o1uor9qk
+[alice]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.17sl3t5a8nu
+[bob]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.mecveaw12b4d
+[carol]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#heading=h.4i5imqvwxv3a
+[testing_commons_notes]: https://docs.google.com/document/d/1TOC8vnmlkWw6HRNHoe5xSv5-qv7LelX6XK3UVCHuwb0/edit#
+[the_doc]: https://docs.google.com/document/d/13bMjmWpsdkgbY-JayrcU-e_QNwRJCP-rHjtqdeeoQHo/edit#
+
+## Motivation
+
+With this proposal we would like to address these three points of friction:
+- For Carol, we introduce an abstraction which is intended to unify existing
```

Comment:
is there something special about a test cluster versus any other cluster?

Reply:
One example: A partial cluster (running only the apiserver & etcd) might be perfectly fine for a certain type of tests.

Reply:
@derekwaynecarr - Also (N) nodes on a single host is another example.  Similar to other DIND clusters.

Reply:
Ability to specify cluster and node type for a suite of tests with specific requirements 
 is a huge advantage. If the abstraction is provider agnostic even better.